[Amos Tan] iP

atyhamos opened the pull request:
# Duchess
> I am looking forward to helping you as much as I can - Duchess of Cambridge, Kate Middleton 👑 [(source)](https://abcnews.go.com/Entertainment/prince-william-kate-middleton-interview-transcript/story?id=12163826) 

Duchess is a Chatbot counterpart based off the Java mascot, [_Duke_](https://www.oracle.com/java/duke.html).
It's
* Text-based
* Easy to learn
* ~~Fast~~ SUPER fast to use

Steps to install:
1. Download it from [here](https://github.com/atyhamos/ip/releases) for **FREE**.
2. Double-click it.
3. Add your tasks.
4. Let her manage your tasks for you😉

Features:
- [x] Todos
- [x] Deadlines
- [x] Events
- [ ] GUI 

***
If you are a Java programmer, you can use it as a template to build upon. Here's the `main` method:

```java
public class Duchess {
    public static void main(String[] args) {
        new Duchess().run();
    }
}
```

atyhamos committed:
Add Level 1. Greet, Echo, Exit

atyhamos committed:
Add Level 2. Add, List

Add documentation
Add DukeList class to store user texts
Change Duke implementation by implementing more methods

atyhamos committed:
Add Level 3. Mark as Done

Add new class, Task, to represent tasks
Add documentation for Task class
Expand DukeList methods implementation

atyhamos committed:
Add Level-4. ToDos, Events, Deadlines

Add new class, ToDo, to specify a task with no date and time
Add new class, Deadline, to specify a task to be done before a specific time
Add new class, Event, to specify a task with a specific start and end time
Modified DukeList, Duke, Task, classes to integrate inheritance of 3 new classes from the Task class

atyhamos committed:
Add A-TextUiTesting

Edit text-ui-test scripts to automate testing of Duke

atyhamos committed:
Add Level-5. Handle Errors

Change Duke to Duchess in file names for more accurate naming
Add new class, DuchessException, to handle exceptions related to Duchess
Update test scripts to handle more specific error checking

atyhamos committed:
Add Level 6. Delete

Add deletion of tasks to Duchess
Change array implementation in DuchessList to ArrayList<Task>
Update test scripts to test for deletion

atyhamos committed:
Add Enums

Use enumerations in Duchess to encapsulate the commands used by Duchess

atyhamos committed:
Add Level-7. Save

Duchess does not store a user's tasks after execution.

This behaviour is unideal as it limits Duchess' functionality.

Let's add Save and Load functionality to Duchess to allow her to
remember user's tasks after exiting from the interface. Make use
of existing Java libraries (File, FileWriter) to implement this.

atyhamos committed:
Add Level 8. Dates and Times.

Duchess stores dates as string.

This behaviour is unideal as it limits Duchess' functionality.

Let's change the string implementation to a LocalDateTime one so
Duchess can become more useful, providing more specific date formats
and performing more calendar type operations.

atyhamos committed:
Merge branch 'branch-Level-7'

atyhamos committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duchess.java
#	src/main/java/Event.java
#	src/main/java/ToDo.java

atyhamos committed:
Add A-MoreOOP: Use More OOP.

Logic for numerous different classes are handled by Duchess.

This is unfavourable as it does not follow the OOP paradigm.

Let's create separate classes Command, Ui, and Parser to handle
their own separate logic apart from Duchess, so Duchess can have
more succinct and readable code.

atyhamos committed:
Add A-Packges. Organize into Packages.

Duchess's classes are not in any package.

This is undesirable as the project grows as the number of classes
would make it messy and disorganized.

Let's organize the classes into packages meaningfully labelled with
their own purposes.

atyhamos committed:
Add A-JUnit: Add JUnit Tests.

Duchess uses I/O redirection to test.

This is undesirable as the project grows and the texts become
more complex and difficult to compare with the expected output.

Let's implement JUnit Tests to test the behaviour of the code.
Let's add Test classes for the different classes within Duchess and
package them into respective packages.

atyhamos committed:
Add A-Jar. Create a JAR File

Duchess does not have an executable JAR file to be distributed
easily.

This is undesirable as it limits the functionality and portability
of Duchess.

Let's package the app as an executable JAR file and release it to
GitHub for others to use.

atyhamos committed:
Add A-JavaDoc.

Duchess' codebase is extensive and will continue to grow as
functionality grows.

Let's add JavaDoc comments to improve readability of code.

atyhamos committed:
Add A-CodingStandard: Follow the Coding Standard

atyhamos committed:
Add Level-9. Find

Duchess stores a list of tasks from the user to display using list.

This can be further improved by limiting the tasks shown using a
search filter.

Let's implement a Find feature to only list out tasks which have
a certain keyword.

atyhamos committed:
Merge branch 'branch-A-CodingStandard'

atyhamos committed:
Merge branch 'branch-Level-9'

atyhamos committed:
Update README.md

atyhamos committed:
Update README.md

atyhamos committed:
Merge branch 'add-gradle-support' of https://github.com/atyhamos/ip
into branch-A-Gradle

yucheng11122017 commented on src/main/java/duchess/command/ByeCommand.java:
```diff
@@ -0,0 +1,30 @@
+ * @author Amos Tan
+ * @version CS2103T AY21/22 Semester 1
+ */
+
```
Shouldn't this blank line be deleted? 😄 I think this may appear in other areas also.

yucheng11122017 commented on src/main/java/duchess/command/Command.java:
```diff
@@ -0,0 +1,58 @@
+public abstract class Command {
+    /** The name of the Command.*/
+    private String name;
+    /** Constructs a Command.*/
```
Shouldn't you include @return and @params for your javadocs? Noticed this is missing for other functions also

yucheng11122017 commented on src/main/java/duchess/command/DeleteCommand.java:
```diff
@@ -0,0 +1,53 @@
+                // "delete" followed by an integer outside of range of the list
+                throw new DuchessException("Apologies, that task does not exist and cannot be deleted.");
+            }
+        } catch (NumberFormatException|DuchessException e) {
```
Shouldn't there be space before and after |? 😄

yucheng11122017 commented on src/main/java/duchess/command/FindCommand.java:
```diff
@@ -0,0 +1,33 @@
+
+public class FindCommand extends Command {
+
+    public FindCommand(String name) {
```
You may be missing a javadoc here and for FindCommand class😢

yucheng11122017 commented on src/main/java/duchess/command/TasksCommand.java:
```diff
@@ -0,0 +1,56 @@
+import duchess.task.Task;
+import java.time.LocalDateTime;
+
+public class TasksCommand extends Command {
```
You may be missing javadoc here also 😢

yucheng11122017 commented on src/main/java/duchess/command/TasksCommand.java:
```diff
@@ -0,0 +1,56 @@
+            duchess.getUi().prettyPrint(tasksToPrint.isBlank() ? "You have no tasks " + keyword + " " + date
+                    : tasksToPrint);
+        } catch (ArrayIndexOutOfBoundsException|DuchessException e) {
+            if (e instanceof  ArrayIndexOutOfBoundsException)
```
Shouldn't you include curly braces here for the if and else statements?

yucheng11122017 commented on src/main/java/duchess/main/Duchess.java:
```diff
@@ -0,0 +1,66 @@
+     * Constructs a Duchess object.
+     */
+    public Duchess()
+    {
```
Shouldn't you be using the Eygptian braces here?

yucheng11122017 commented on src/main/java/duchess/main/Duchess.java:
```diff
@@ -0,0 +1,66 @@
+        ui.prettyPrint("Good day. I am Duchess.\nWhat can I do for you?");
+    }
+
+    public static void main(String[] args) {
```
You might be missing a javadoc here.

yucheng11122017 commented on src/main/java/duchess/main/DuchessFileHandler.java:
```diff
@@ -0,0 +1,139 @@
+
+    /**
+     * Extracts the DuchessList from the save file.
+     * @param existingDuchessFile The save file.
```
Shouldn't you include a empty line between description and parameter? I think this may come out in other areas also

yucheng11122017 commented on src/main/java/duchess/main/DuchessList.java:
```diff
@@ -0,0 +1,90 @@
+ * @version CS2103T AY21/22 Semester 1
+ */
+
+public class DuchessList {
```
I think you may have forgotten to use Egyptian braces for this class 😢

yucheng11122017 commented on src/main/java/duchess/main/DuchessList.java:
```diff
@@ -0,0 +1,90 @@
+     */
+    public Task getTask(int listNumber) throws IllegalArgumentException
+    {
+        if (checkWithinRange(listNumber))
```
Don't forget the curly brackets around your if statement 😄

yucheng11122017 commented on src/main/java/duchess/main/Ui.java:
```diff
@@ -0,0 +1,31 @@
+     * Prints fancily including horizontal bars on top and bottom.
+     * @param input String to be printed fancily.
+     */
+    public void prettyPrint(String input)
```
Shouldn't you use Eygptian braces here and for the one below?

yucheng11122017 commented on src/main/java/duchess/task/Event.java:
```diff
@@ -0,0 +1,103 @@
+    public String toString() {
+        return "[E]" + super.toString()
+                + " (at: " + dateTimeStart.format(PRINT_DATE_FORMATTER) + " to "
+                    + dateTimeEnd.format(PRINT_DATE_FORMATTER) + ")";
```
Shouldn't you be wrapping with 8 spaces instead? 🤔

yucheng11122017 commented on src/test/java/duchess/command/ByeCommandTest.java:
```diff
@@ -0,0 +1,16 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class ByeCommandTest {
```
Shouldn't you include javadoc here?

yucheng11122017 reviewed (commented):
Hi Amos, I think your code is pretty cool. I quite liked the enum for commands that you did. 

I think the main issues are
- The Eygptian braces 
- Missing some javadocs here and there
Apart from that I think its ok! (I hope I didn't miss anything) 
Hope you find this useful!

atyhamos committed:
Update README.md

atyhamos committed:
Add Level-10. GUI

Duchess uses command line interface to display text feedback to the
user.

This can be further improved by adding a graphical user interface
(GUI) to make it more customisable.

Let's implement a GUI for Duchess to make it look and feel more
modern.

atyhamos committed:
Merge branch 'branch-Level-10'

tanjoewel commented on src/main/java/duchess/command/Command.java:
```diff
@@ -0,0 +1,58 @@
+        return null;
+    }
+
+    public String getName() {
+        return this.name;
+    }
```
Perhaps the method name here could be more descriptive? I saw this method used in the child classes and was slightly confused as to what it was getting the name of. Maybe getNameOfCommand would be a clearer method name?

atyhamos replied:
Agreed. It is ambiguous. I've changed it to getCommandName(). Thanks for the suggestion.

tanjoewel commented on src/main/java/duchess/task/Event.java:
```diff
@@ -0,0 +1,103 @@
+     */
+    public static LocalDateTime[] convertStringToDate (String date, String duration) throws DuchessException {
+        try {
+            String replacement = duration.replace("am", "AM").replace("pm","PM");
```
Should there be a white space between the comma and "PM"?

tanjoewel commented on src/main/java/duchess/task/Event.java:
```diff
@@ -0,0 +1,103 @@
+        }
+    }
+
+    public LocalDateTime getDateTime(){
+        return this.dateTimeStart;
+    }
+
+    /**
+     * Returns a simplified representation of the Event for easier recovery from save file.
+     * @return The file formatted string representation of the Event.
+     */
+    public String toFileFormat() {
+        return String.format("E%s,%s,%s,%s", name, dateTimeStart.format(PRINT_DATE_FORMATTER),
+                dateTimeEnd.format(PRINT_DATE_FORMATTER), isDone);
+    }
+
+    /**
+     * Converts the user input string to a LocalDateTime.
+     * @param by The date as text within the file.
+     * @return The LocalDateTime representation.
+     * @throws DuchessException Exception thrown when an incorrect format is used for Deadline.
+     */
+    public static LocalDateTime convertTextToDate (String by)  {
+        return LocalDateTime.parse(by, PRINT_DATE_FORMATTER);
+    }
```
Should there just be a single space between the method name and the curly bracket? Line 70 there is no space and line 89 there are two spaces
I also noticed this in your other child Task classes

tanjoewel commented on src/main/java/duchess/task/Task.java:
```diff
@@ -0,0 +1,69 @@
+        return name.contains(keyword);
+    }
+
+    public boolean isDone() {
+        return isDone;
+    }
+
+    public abstract LocalDateTime getDateTime();
```
Should there be JavaDocs for these two methods?

atyhamos replied:
I generally avoided putting JavaDocs for getters (and setters) as per the coding standard.
_"You MUST write header comments for all classes, public methods except for getters/setters."_
I see how isDone() may not look like a getter method, so I've changed it to getStatus(), but getDateTime() should be clear.

tanjoewel commented on src/main/java/duchess/task/ToDo.java:
```diff
@@ -0,0 +1,42 @@
+        return LocalDateTime.MAX;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[T]" + super.toString());
+    }
+}
```
Should there be JavaDocs for this method?

tanjoewel reviewed (commented):
Hi Amos! LGTM! Most of the coding standard violations have been mentioned by @yucheng11122017 but do take note of the Egyptian Braces and white spaces between name and curly bracket as well as things like "+" and variable name!
Other than that I like how the code is clearly written and easy to navigate, as well as your use of comments which really helped to explain what the idea is behind certain lines of code! Good job!

atyhamos committed:
Add CheckStyle

atyhamos committed:
Merge branch 'branch-A-CheckStyle'

atyhamos committed:
Improve GUI

atyhamos committed:
Improve code quality.

With the implementation of GUI, old UI code has become obsolete.
Command class' methods are also are quite lengthy.

This can be rectified by using refactoring and cleaning up messy
code.

Let's use refactoring and make methods clearer in the Command class'
methods. Let's also remove the old Ui class and I/O test files which
are obsolete with GUI and JUnit.

atyhamos committed:
Merge pull request #1 from atyhamos/branch-A-CodeQuality

Improve code quality.

atyhamos committed:
Merge branch 'master' of https://github.com/atyhamos/ip

atyhamos committed:
Add Assertions.

Code in Duchess makes assumptions about behaviour in methods and lets
the execution continue until an exception is thrown.

This can be improved by clarifying assumptions made in code so assertion
failures can be detected.

Let's implement Assertions to Duchess' methods to define assumptions
made about the program state.

atyhamos committed:
Merge branch 'master' into branch-A-Assertions

atyhamos committed:
Merge pull request #2 from atyhamos/branch-A-Assertions

Add Assertions.

atyhamos committed:
Merge branch 'master' of https://github.com/atyhamos/ip

atyhamos committed:
Add duplicate detection

Duchess allows for duplicate tasks in the list to be stored.

This is unideal as unnecessary repetitions may clutter the list and
would confuse the user.

Let's add duplicate detection to only allow unique todos, deadlines,
and events to be added to the list.

atyhamos committed:
Improve GUI

atyhamos committed:
Set theme jekyll-theme-modernist

atyhamos committed:
Update README.md

atyhamos committed:
Add Ui image to docs

atyhamos committed:
Merge branch 'master' of https://github.com/atyhamos/ip

atyhamos committed:
Update user profile picture to generic template
